worker/uniter/container/resolver.go

```diff
@@ -0,0 +1,71 @@
+// Copyright 2020 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package container
+
+import (
+	"github.com/juju/loggo"
+
+	"github.com/juju/juju/worker/uniter/operation"
+	"github.com/juju/juju/worker/uniter/remotestate"
+	"github.com/juju/juju/worker/uniter/resolver"
+)
+
+var logger = loggo.GetLogger("juju.worker.uniter.actions")
+
+type containerResolver struct{}
+
+// NewResolver returns a new resolver with determines which action related operation
+// should be run based on local and remote uniter states.
+func NewResolver() resolver.Resolver {
+	return &containerResolver{}
+}
+
+// NextOp implements the resolver.Resolver interface.
+func (r *containerResolver) NextOp(
+	localState resolver.LocalState,
+	remoteState remotestate.Snapshot,
+	opFactory operation.Factory,
+) (operation.Operation, error) {
+	noOp := func() (operation.Operation, error) {
+		if localState.Kind == operation.RemoteInit {
+			// If we are resuming from an unexpected state, skip init.
+			// Retry will occur when remotestate updates.
```

Comment:
Is this deterministic? In theory the uniter could block?

Reply:
Uniter can't block because we need to be able to run hooks. We only block actions atm.